File: go/internal/monorepo/workspace_manager_test.go
package monorepo_test

import (
	"testing"

	"github.com/microsoft/beachball/internal/monorepo"
	"github.com/microsoft/beachball/internal/options"
	"github.com/microsoft/beachball/internal/testutil"
	"github.com/microsoft/beachball/internal/types"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func getTestOptions(rootPath string) types.BeachballOptions {
	opts := types.DefaultOptions()
	opts.Path = rootPath
	opts.Branch = testutil.DefaultRemoteBranch
	opts.Fetch = false
	return opts
}

// TS: pnpm workspace detection
func TestGetPackageInfos_PnpmWorkspace(t *testing.T) {
	// Create a monorepo with packages but no npm workspaces in package.json
	rootPkg := map[string]any{
		"name":    "pnpm-monorepo",
		"version": "1.0.0",
		"private": true,
	}
	packages := map[string]map[string]any{
		"foo": {"name": "foo", "version": "1.0.0"},
		"bar": {"name": "bar", "version": "1.0.0"},
	}
	factory := testutil.NewCustomRepositoryFactory(t, rootPkg, map[string]map[string]map[string]any{
		"packages": packages,
	})
	repo := factory.CloneRepository()

	// Add pnpm-workspace.yaml
	repo.WriteFileContent("pnpm-workspace.yaml", "packages:\n  - 'packages/*'\n")
	repo.Git([]string{"add", "-A"})
	repo.Git([]string{"commit", "-m", "add pnpm config"})

	opts := getTestOptions(repo.RootPath())
	parsed := options.GetParsedOptionsForTest(repo.RootPath(), types.CliOptions{}, opts)
	infos, err := monorepo.GetPackageInfos(&parsed.Options)
	require.NoError(t, err)

	assert.Contains(t, infos, "foo")
	assert.Contains(t, infos, "bar")
}

// TS: lerna workspace detection
func TestGetPackageInfos_LernaWorkspace(t *testing.T) {
	rootPkg := map[string]any{
		"name":    "lerna-monorepo",
		"version": "1.0.0",
		"private": true,
	}
	packages := map[string]map[string]any{
		"foo": {"name": "foo", "version": "1.0.0"},
		"bar": {"name": "bar", "version": "1.0.0"},
	}
	factory := testutil.NewCustomRepositoryFactory(t, rootPkg, map[string]map[string]map[string]any{
		"packages": packages,
	})
	repo := factory.CloneRepository()

	// Add lerna.json
	repo.WriteFileContent("lerna.json", `{"packages": ["packages/*"]}`)
	repo.Git([]string{"add", "-A"})
	repo.Git([]string{"commit", "-m", "add lerna config"})

	opts := getTestOptions(repo.RootPath())
	parsed := options.GetParsedOptionsForTest(repo.RootPath(), types.CliOptions{}, opts)
	infos, err := monorepo.GetPackageInfos(&parsed.Options)
	require.NoError(t, err)

	assert.Contains(t, infos, "foo")
	assert.Contains(t, infos, "bar")
}

// TS: rush workspace detection
func TestGetPackageInfos_RushWorkspace(t *testing.T) {
	rootPkg := map[string]any{
		"name":    "rush-monorepo",
		"version": "1.0.0",
		"private": true,
	}
	packages := map[string]map[string]any{
		"foo": {"name": "foo", "version": "1.0.0"},
		"bar": {"name": "bar", "version": "1.0.0"},
	}
	factory := testutil.NewCustomRepositoryFactory(t, rootPkg, map[string]map[string]map[string]any{
		"packages": packages,
	})
	repo := factory.CloneRepository()

	// Add rush.json with project folders
	repo.WriteFileContent("rush.json", `{
		"projects": [
			{"packageName": "foo", "projectFolder": "packages/foo"},
			{"packageName": "bar", "projectFolder": "packages/bar"}
		]
	}`)
	repo.Git([]string{"add", "-A"})
	repo.Git([]string{"commit", "-m", "add rush config"})

	opts := getTestOptions(repo.RootPath())
	parsed := options.GetParsedOptionsForTest(repo.RootPath(), types.CliOptions{}, opts)
	infos, err := monorepo.GetPackageInfos(&parsed.Options)
	require.NoError(t, err)

	assert.Contains(t, infos, "foo")
	assert.Contains(t, infos, "bar")
}

File: rust/tests/workspace_manager_test.rs
mod common;

use beachball::monorepo::package_infos::get_package_infos;
use beachball::options::get_options::get_parsed_options_for_test;
use beachball::types::options::{BeachballOptions, CliOptions};
use common::DEFAULT_REMOTE_BRANCH;
use common::repository_factory::RepositoryFactory;
use serde_json::json;
use std::collections::HashMap;

fn make_opts() -> BeachballOptions {
    BeachballOptions {
        branch: DEFAULT_REMOTE_BRANCH.to_string(),
        fetch: false,
        ..Default::default()
    }
}

#[test]
fn pnpm_workspace_detection() {
    let root = json!({
        "name": "pnpm-monorepo",
        "version": "1.0.0",
        "private": true
    });
    let packages = HashMap::from([
        (
            "foo".to_string(),
            json!({"name": "foo", "version": "1.0.0"}),
        ),
        (
            "bar".to_string(),
            json!({"name": "bar", "version": "1.0.0"}),
        ),
    ]);
    let factory = RepositoryFactory::new_custom(root, vec![("packages".to_string(), packages)]);
    let repo = factory.clone_repository();

    // Remove yarn.lock (added by factory) and add pnpm-workspace.yaml
    repo.git(&["rm", "yarn.lock"]);
    repo.write_file_content("pnpm-workspace.yaml", "packages:\n  - 'packages/*'\n");
    repo.git(&["add", "-A"]);
    repo.git(&["commit", "-m", "switch to pnpm"]);

    let parsed = get_parsed_options_for_test(repo.root_path(), CliOptions::default(), make_opts());
    let infos = get_package_infos(&parsed.options).unwrap();

    assert!(infos.contains_key("foo"), "expected foo in {infos:?}");
    assert!(infos.contains_key("bar"), "expected bar in {infos:?}");
}

#[test]
fn lerna_workspace_detection() {
    let root = json!({
        "name": "lerna-monorepo",
        "version": "1.0.0",
        "private": true
    });
    let packages = HashMap::from([
        (
            "foo".to_string(),
            json!({"name": "foo", "version": "1.0.0"}),
        ),
        (
            "bar".to_string(),
            json!({"name": "bar", "version": "1.0.0"}),
        ),
    ]);
    let factory = RepositoryFactory::new_custom(root, vec![("packages".to_string(), packages)]);
    let repo = factory.clone_repository();

    // Add lerna.json
    repo.write_file_content("lerna.json", r#"{"packages": ["packages/*"]}"#);
    repo.git(&["add", "-A"]);
    repo.git(&["commit", "-m", "add lerna config"]);

    let parsed = get_parsed_options_for_test(repo.root_path(), CliOptions::default(), make_opts());
    let infos = get_package_infos(&parsed.options).unwrap();

    assert!(infos.contains_key("foo"), "expected foo in {infos:?}");
    assert!(infos.contains_key("bar"), "expected bar in {infos:?}");
}

#[test]
fn rush_workspace_detection() {
    let root = json!({
        "name": "rush-monorepo",
        "version": "1.0.0",
        "private": true
    });
    let packages = HashMap::from([
        (
            "foo".to_string(),
            json!({"name": "foo", "version": "1.0.0"}),
        ),
        (
            "bar".to_string(),
            json!({"name": "bar", "version": "1.0.0"}),
        ),
    ]);
    let factory = RepositoryFactory::new_custom(root, vec![("packages".to_string(), packages)]);
    let repo = factory.clone_repository();

    // Remove yarn.lock (added by factory) and add rush.json
    repo.git(&["rm", "yarn.lock"]);
    repo.write_file_content(
        "rush.json",
        r#"{
        "projects": [
            {"packageName": "foo", "projectFolder": "packages/foo"},
            {"packageName": "bar", "projectFolder": "packages/bar"}
        ]
    }"#,
    );
    repo.git(&["add", "-A"]);
    repo.git(&["commit", "-m", "add rush config"]);

    let parsed = get_parsed_options_for_test(repo.root_path(), CliOptions::default(), make_opts());
    let infos = get_package_infos(&parsed.options).unwrap();

    assert!(infos.contains_key("foo"), "expected foo in {infos:?}");
    assert!(infos.contains_key("bar"), "expected bar in {infos:?}");
}
